platform/cervantes/switch.sh

```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+# This script exits KOReader and loads the default BQ Reader App
+# Copy it to /usr/bin so that it can be used by simply writing switch.sh on the terminal emulator plugin of KOReader
+#
+# It can also be invoked with the exit command on KOReader by modifying /etc/rc.local like this:
+#
+# REPLACE THIS CODE:
+# check if KOReader script exists.
+#  if [ -x /mnt/private/koreader/koreader.sh ]; then
+#     # yada! KOReader is installed and ready to run.
+#     while true; do
+#         /mnt/private/koreader/koreader.sh
+#         if [ -x /usr/bin/safemode ]; then
+#             safemode storage || sleep 1
+#         else
+#             sleep 1
+#         fi
+#     done
+#
+#
+# WITH THIS CODE:
+# check if KOReader script exists.
+#  if [ -x /mnt/private/koreader/koreader.sh ]; then
+#     # yada! KOReader is installed and ready to run.
+#     while true; do
+#         /mnt/private/koreader/koreader.sh
+#         /usr/bin/switch.sh
+#     done
+
+killall koreader.sh
+killall reader.lua        # Exit KOReader so that the touchcreen can be used by BQ Reader App
```

Comment:
As an aside, I don't believe we handle SIGTERM (or any of those types of signals).

Reply:
Well I'm running KOReader by default via koreader-standalone.sh as rc.local, so that way when I exit KOReader nothing happens, it just restarts KOReader again. With this script I can change to BQ Reader whenever i need it by issuing a command on the terminal (simple way to do it if the user for whatever reason doesn't want to mess with the code on koreader-standalone.sh) and if I enable it the way I suggest on rc.local, I can make it switch with the exit option (which makes it useful for something, as right now it just restarts KOReader and there is an option already there for that). 

Futhermore, I don't know how am I supposed to execute /mnt/private/koreader/koreader.sh from BQ reader, there is no way to execute scripts from BQ Reader as far as I know, and if there is, exiting BQ Reader and opening KOReader would be as simple as running stopapp.sh and the koreader.sh. Right now, if I want to switch from BQ Reader back to KOReader I just power off and then on again the device, and that loads KOReader back

I made this because right now because of a bug I don't have wifi access on KOReader, and I needed a way to switch to BQ Reader in order to use wifi and access my ebook via telnet. I started tinkering with the device and came to this solution, and as I am no developer and I use to hack my way around things I'm sure there is a better way of doing this (that is regarding to Frenzie's comment on SIGTERM)

On a side note, I noticed that the Store button on BQ's dev firmware doesn't work (I believe it is on purpose because on licenses and all that jazz) so maybe it could be modified to lauch KOReader?

Reply:
> Well I'm running KOReader by default via koreader-standalone.sh as rc.local, so that way when I exit KOReader nothing happens, it just restarts KOReader again

> I don't know how am I supposed to execute /mnt/private/koreader/koreader.sh from BQ reader, there is no way to execute scripts from BQ Reader as far as I know

Both things are explained in https://www.mobileread.com/forums/showpost.php?p=3760905&postcount=2. I didn't receive any feedback about the instructions. I'm more than happy to adapt instructions to be easier to read/understand :smile: 

The reason why it is not possible to launch stuff from BQ Reader is explained on https://www.mobileread.com/forums/showpost.php?p=3760904&postcount=1

Now I know that the buggy fw only affects Cervantes3/4 and no older models. So porting fmon/kfmon should be easy since inotify is enabled inside the kernel.

> With this script I can change to BQ Reader whenever i need it by issuing a command on the terminal (simple way to do it if the user for whatever reason doesn't want to mess with the code on koreader-standalone.sh) and if I enable it the way I suggest on rc.local, I can make it switch with the exit option (which makes it useful for something, as right now it just restarts KOReader and there is an option already there for that).

IMO changes that affect the interaction between KOReader and stock software should stay in koreader.sh and koreader-standalone.sh is just a wrapper to run koreader.sh on a device without relying on other system scripts.

> I needed a way to switch to BQ Reader in order to use wifi and access my ebook via telnet. 

Not entirely related, but you can always start telnet at boot (if you're executing koreader-standalone.sh as rc.local) pressing the home button while the little dots are on screen. From shell you can start both KOReader and BQ reader without problems (because both programs lock the home button while they are running). Once you exited from the program pressing the home button again will continue normal boot and start KOReader as usual.

> On a side note, I noticed that the Store button on BQ's dev firmware doesn't work (I believe it is on purpose because on licenses and all that jazz) so maybe it could be modified to lauch KOReader?

Not without adapting and rebuilding the Qt application. Porting fmon seems easier, IDK. Pinging @avsej which did fix a bug before on bq software.

Reply:
> Both things are explained in https://www.mobileread.com/forums/showpost.php?p=3760905&postcount=2. I didn't receive any feedback about the instructions. I'm more than happy to adapt instructions to be easier to read/understand

I already know that I can launch scripts over telnet, I meant there is no way to launch scripts inside BQ Reader App in a standalone way, that is without depending on another device to launch them via command line

> Not entirely related, but you can always start telnet at boot (if you're executing koreader-standalone.sh as rc.local) pressing the home button while the little dots are on screen. From shell you can start both KOReader and BQ reader without problems (because both programs lock the home button while they are running). Once you exited from the program pressing the home button again will continue normal boot and start KOReader as usual.

This right now is useless since networking is broken on the Cervantes 2013 via KOReader

> Not without adapting and rebuilding the Qt application. Porting fmon seems easier, IDK. Pinging @avsej which did fix a bug before on bq software

Whatever works best or is easiest to implement, if it gets the job done, then it is enough for me